Họ Tên: Nguyễn Thị Mến
MSSV: 20424105

In [22]:
import pandas as pd
import numpy as np
import random as rd

In [23]:
pd.set_option('display.max_columns', 10)

train_data = pd.read_csv("train.csv", encoding='utf-8')

print(train_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None


In [24]:
print(train_data.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


Theo như dữ liệu thì ta thấy thuộc tính Cabin có 204 dòng, Age có 714 dòng, Embarked có 889 dòng. Các thuộc tính này hiện đang bị khuyết. 
Bây giờ chúng ta thực hiện việc xóa các cột không ảnh hướng đến việc phân tích:

In [25]:
train_data.drop(['PassengerId', 'Ticket', 'Cabin'], axis = 1, inplace = True)

Xuất thông tin các cột còn lại.

In [26]:
train_data.iloc[0:10]

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   
5         0       3                                   Moran, Mr. James   
6         0       1                            McCarthy, Mr. Timothy J   
7         0       3                     Palsson, Master. Gosta Leonard   
8         1       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9         1       2                Nasser, Mrs. Nicholas (Adele Achem)   

      Sex   Age  SibSp  Parch     Fare Embarked  
0    male  22.0      1      0   7.2500        S  
1  female  38.0      1      0  71.2833        C  
2  female  26.0      0      0   7.9250        S  
3  female  35.0      1      0  53.1000        S  
4    male  35.0      0      0   8.0500        S  
5    male   NaN      0      0   8.4583        Q  
6    male  54.0      0      0  51.8625        S  
7    male   2.0      3      1  21.0750        S  
8  female  27.0      0      2  11.1333        S  
9  female  14.0      1      0  30.0708        C

Tiếp theo ta xóa dữ liệu khuyết.

In [27]:
train_data['Age'].fillna(train_data['Age'].median(), inplace = True)

train_data['Embarked'].fillna(train_data['Embarked'].mode()[0], inplace = True)

train_data['Fare'].fillna(train_data['Fare'].median(), inplace = True)

In [28]:
train_data.iloc[0:10]

Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   
5         0       3                                   Moran, Mr. James   
6         0       1                            McCarthy, Mr. Timothy J   
7         0       3                     Palsson, Master. Gosta Leonard   
8         1       3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9         1       2                Nasser, Mrs. Nicholas (Adele Achem)   

      Sex   Age  SibSp  Parch     Fare Embarked  
0    male  22.0      1      0   7.2500        S  
1  female  38.0      1      0  71.2833        C  
2  female  26.0      0      0   7.9250        S  
3  female  35.0      1      0  53.1000        S  
4    male  35.0      0      0   8.0500        S  
5    male  28.0      0      0   8.4583        Q  
6    male  54.0      0      0  51.8625        S  
7    male   2.0      3      1  21.0750        S  
8  female  27.0      0      2  11.1333        S  
9  female  14.0      1      0  30.0708        C

In [29]:
train_data['People_Number'] = train_data['SibSp'] + train_data['Parch'] + 1

train_data['Name_c'] = train_data['Name'].str.split(", ", expand=True)[1].str.split(".", expand=True)[0]

train_data['Fare_c'] = pd.qcut(train_data['Fare'], 4)

train_data['Age_c'] = pd.cut(train_data['Age'].astype(int), 5)

print(train_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   Survived       891 non-null    int64   
 1   Pclass         891 non-null    int64   
 2   Name           891 non-null    object  
 3   Sex            891 non-null    object  
 4   Age            891 non-null    float64 
 5   SibSp          891 non-null    int64   
 6   Parch          891 non-null    int64   
 7   Fare           891 non-null    float64 
 8   Embarked       891 non-null    object  
 9   People_Number  891 non-null    int64   
 10  Name_c         891 non-null    object  
 11  Fare_c         891 non-null    category
 12  Age_c          891 non-null    category
dtypes: category(2), float64(2), int64(5), object(4)
memory usage: 78.9+ KB
None


In [30]:
train_data.iloc[0:0,10]

Series([], Name: Name_c, dtype: object)

Chúng ta sẽ quan tâm đến hai đặc trưng là Sex và Embarked. Đầu tiên ta sẽ biến đổi các thuộc tính này thành biểu diễn dạng số học.

In [31]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

name_le = LabelEncoder()
name_labels = name_le.fit_transform(train_data['Name_c'].astype(str))
train_data['Name_Label'] = name_labels

sex_le = LabelEncoder()
sex_labels = sex_le.fit_transform(train_data['Sex'].astype(str))
train_data['Sex_Label'] = sex_labels

fare_le = LabelEncoder()
fare_labels = fare_le.fit_transform(train_data['Fare_c'].astype(str))
train_data['Fare_Label'] = fare_labels

age_le = LabelEncoder()
age_labels = age_le.fit_transform(train_data['Age_c'].astype(str))
train_data['Age_Label'] = age_labels

emb_le = LabelEncoder()
emb_labels = emb_le.fit_transform(train_data['Embarked'].astype(str))
train_data['Embarked_Label'] = emb_labels

poke_df_sub = train_data[['Survived', 'Pclass', 'Name_Label', 
                            'Sex_Label', 'Age_Label', 'SibSp', 
                            'Parch', 'Fare_Label', 'Embarked_Label', 
                            'People_Number']]
print(poke_df_sub.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   Survived        891 non-null    int64
 1   Pclass          891 non-null    int64
 2   Name_Label      891 non-null    int32
 3   Sex_Label       891 non-null    int32
 4   Age_Label       891 non-null    int32
 5   SibSp           891 non-null    int64
 6   Parch           891 non-null    int64
 7   Fare_Label      891 non-null    int32
 8   Embarked_Label  891 non-null    int32
 9   People_Number   891 non-null    int64
dtypes: int32(5), int64(5)
memory usage: 52.3 KB
None


Sau khi biểu diễn xong ta các cột dưới đây mô tả dưới dạng số học. Tiếp theo chúng ta sẽ sử dụng One-hot Encoding cho các đặc trưng này.

In [32]:
#Name
name_ohe = OneHotEncoder()
name_feature_arr = name_ohe.fit_transform(
                                poke_df_sub[['Name_Label']]).toarray()
name_feature_labels = ['Name_' + str(cls_label) 
                           for cls_label in name_le.classes_]
name_features = pd.DataFrame(name_feature_arr, 
                            columns = name_feature_labels)
#Sex
sex_ohe = OneHotEncoder()
sex_feature_arr = sex_ohe.fit_transform(
                                poke_df_sub[['Sex_Label']]).toarray()
sex_feature_labels = ['Sex_' + str(cls_label) 
                           for cls_label in sex_le.classes_]
sex_features = pd.DataFrame(sex_feature_arr, 
                            columns = sex_feature_labels)
#Fare
fare_ohe = OneHotEncoder()
fare_feature_arr = fare_ohe.fit_transform(
                                poke_df_sub[['Fare_Label']]).toarray()
fare_feature_labels = ['Fare_' + str(cls_label) 
                           for cls_label in fare_le.classes_]
fare_features = pd.DataFrame(fare_feature_arr, 
                            columns = fare_feature_labels)
#Age
age_ohe = OneHotEncoder()
age_feature_arr = age_ohe.fit_transform(
                                poke_df_sub[['Age_Label']]).toarray()
age_feature_labels = ['Age_' + str(cls_label) 
                           for cls_label in age_le.classes_]
age_features = pd.DataFrame(age_feature_arr, 
                            columns = age_feature_labels)

#Embarked
emb_ohe = OneHotEncoder()
emb_feature_arr = emb_ohe.fit_transform(
                                poke_df_sub[['Embarked_Label']]).toarray()
emb_feature_labels = ['Embarked_' + str(cls_label) 
                           for cls_label in emb_le.classes_]
emb_features = pd.DataFrame(emb_feature_arr, 
                            columns = emb_feature_labels)

Bây giờ thì ta gọp các đặc trưng được tạo ra ở trên lại.

In [33]:
poke_df_ohe = pd.concat([poke_df_sub, name_features, 
                         sex_features, fare_features, 
                         age_features, emb_features], axis=1)
# columns = sum([['Survived', 'Pclass'], 
#                name_feature_labels,
#                sex_feature_labels, 
#                fare_feature_labels,
#                age_feature_labels,
#                ['SibSp', 'Parch'],
#                emb_feature_labels], [])

In [34]:
poke_df_ohe.drop(['Name_Label', 'People_Number', 'Sex_Label', 'Age_Label'
                  , 'Embarked_Label', 
                 'Fare_Label'], axis = 1, inplace = True)
print(poke_df_ohe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 35 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Survived              891 non-null    int64  
 1   Pclass                891 non-null    int64  
 2   SibSp                 891 non-null    int64  
 3   Parch                 891 non-null    int64  
 4   Name_Capt             891 non-null    float64
 5   Name_Col              891 non-null    float64
 6   Name_Don              891 non-null    float64
 7   Name_Dr               891 non-null    float64
 8   Name_Jonkheer         891 non-null    float64
 9   Name_Lady             891 non-null    float64
 10  Name_Major            891 non-null    float64
 11  Name_Master           891 non-null    float64
 12  Name_Miss             891 non-null    float64
 13  Name_Mlle             891 non-null    float64
 14  Name_Mme              891 non-null    float64
 15  Name_Mr               8

In [35]:
poke_df_ohe.iloc[1:]

Survived  Pclass  SibSp  Parch  Name_Capt  ...  Age_(48.0, 64.0]  \
1           1       1      1      0        0.0  ...               0.0   
2           1       3      0      0        0.0  ...               0.0   
3           1       1      1      0        0.0  ...               0.0   
4           0       3      0      0        0.0  ...               0.0   
5           0       3      0      0        0.0  ...               0.0   
..        ...     ...    ...    ...        ...  ...               ...   
886         0       2      0      0        0.0  ...               0.0   
887         1       1      0      0        0.0  ...               0.0   
888         0       3      1      2        0.0  ...               0.0   
889         1       1      0      0        0.0  ...               0.0   
890         0       3      0      0        0.0  ...               0.0   

     Age_(64.0, 80.0]  Embarked_C  Embarked_Q  Embarked_S  
1                 0.0         1.0         0.0         0.0  
2                 0.0         0.0         0.0         1.0  
3                 0.0         0.0         0.0         1.0  
4                 0.0         0.0         0.0         1.0  
5                 0.0         0.0         1.0         0.0  
..                ...         ...         ...         ...  
886               0.0         0.0         0.0         1.0  
887               0.0         0.0         0.0         1.0  
888               0.0         0.0         0.0         1.0  
889               0.0         1.0         0.0         0.0  
890               0.0         0.0         1.0         0.0  

[890 rows x 35 columns]

Cho 5 xúc xắc, lắc 1000 lần. Tính tỉ lệ ra mặt 1 cho cả 5 xúc xắc.

In [36]:
def laMatMotDiem():
    return rd.randint(1, 6) == 1

In [37]:
def tungXucXac():
    for i in range(5):
        if laMatMotDiem():
            continue
        return False
    return True

In [38]:
def tungNLan(n):
    dem = 0
    for i in range(n):
        if tungXucXac():
            dem += 1
    return dem/n

In [39]:
print(f'Tỉ lệ tung được 5 xúc xắc mặt 1 là: {tungNLan(1000)}%')

Tỉ lệ tung được 5 xúc xắc mặt 1 là: 0.0%
